cpp/src/compaction/stream_compaction_ops.cu

```diff
@@ -149,185 +87,146 @@ struct bit_or: public thrust::unary_function<thrust::tuple<gdf_valid_type,gdf_va
 		return thrust::get<0>(x) | thrust::get<1>(x);
 	}
 };
- 
-
-typedef thrust::transform_iterator<modulus_bit_width, thrust::counting_iterator<gdf_size_type> > bit_position_iterator;
-
 
-template<typename stencil_type>
-struct is_stencil_true
+struct is_bit_set
 {
 	__host__ __device__
-	bool operator()(const thrust::tuple<stencil_type, gdf_valid_iterator::value_type, bit_position_iterator::value_type> value)
+	bool operator()(const thrust::tuple<gdf_size_type, thrust::device_ptr<gdf_valid_type>> value)
 	{
-		gdf_size_type position = thrust::get<2>(value);
+		gdf_size_type position = thrust::get<0>(value);
 
-		return ((thrust::get<1>(value) >> position) & 1) && (thrust::get<0>(value) != 0);
+		return gdf_is_valid(thrust::get<1>(value).get(), position);
 	}
-};
+}; 
 
-struct is_bit_set
+typedef thrust::tuple<thrust::counting_iterator<gdf_size_type>, thrust::constant_iterator<gdf_valid_type*>> mask_tuple;
+typedef thrust::zip_iterator<mask_tuple> zipped_mask;
+
+typedef thrust::transform_iterator<is_bit_set, zipped_mask> bit_set_iterator;
+
+template<typename stencil_type>
+struct is_stencil_true
 {
 	__host__ __device__
-	bool operator()(const thrust::tuple< gdf_valid_iterator::value_type, bit_position_iterator::value_type> value)
+	bool operator()(const thrust::tuple<stencil_type, bit_set_iterator::value_type> value)
 	{
-		gdf_size_type position = thrust::get<1>(value);
-
-		return ((thrust::get<0>(value) >> position) & 1);
+		return thrust::get<1>(value) && (thrust::get<0>(value) != 0);
 	}
-}; 
+};
 
 struct bit_mask_pack_op : public thrust::unary_function<int64_t,gdf_valid_type>
 {
+	static_assert(sizeof(gdf_valid_type) == 1, "Unexpected size of gdf_valid_type");
 	__host__ __device__
 		gdf_valid_type operator()(const int64_t expanded)
 		{
 			gdf_valid_type result = 0;
-			for(unsigned int i = 0; i < GDF_VALID_BITSIZE; i++){
-				// 0, 8, 16, ....,48,  56
-				unsigned char byte = (expanded >> ( (GDF_VALID_BITSIZE - 1 - i )  * 8));
+			for(unsigned i = 0; i < GDF_VALID_BITSIZE; i++){
+				unsigned char byte = (expanded >> (i * CHAR_BIT));
 				result |= (byte & 1) << i;
 			}
-			return (result);
+			return result;
 		}
 };
 
+//zip the stencil and the valid iterator together
+typedef thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int8_t>>, bit_set_iterator> zipped_stencil_tuple;
+typedef thrust::zip_iterator<zipped_stencil_tuple> zipped_stencil_iterator;
```

Comment:
`using` for type declarations is preferred. This will be among the first things I'd want added to our coding style guide.
```suggestion
using zipped_stencil_iterator = thrust::zip_iterator<zipped_stencil_tuple>;
```

Reply:
Suggested change committed.